### DISCUSSION TITLE
Component: Tag Input, or StringArrayEditor.jsx

### DISCUSSION THREAD

**Discussion by Gazook89** [naturalcrit/homebrewery#3466] [Ideas]
Been looking at the StringArrayEditor component, otherwise known as the Tag Input, and considering options.

**Relevant Issues:**
#2333
#2335 
#3462
#3166
#2357

**Relevant PR:**
#3469

### Keyboard

- [x] `,` | `Enter` to submit tag and move to next input
- [x] Arrow keys to highlight and move between submitted tags
- [x] ArrowLeft from 'new tag' input, if empty, moves to previous submitted tag, if it exists
- [x] `Enter` | `Space` to edit existing tag
- [x] `Escape` to cease editing and not save changes, and focus same tag
- [x] `,` | `Enter` to submit change to existing tag, and focus "new tag" input
- [x] `Delete` to remove an existing focused tag
- [ ] `Tab` to enter, and exit, tag container.  When entering, focus the "new tag" input

### Mouse

- [x] `onMouseUp` on existing tag to edit (single click to start editing)
- [x] `onMouseUp` on **X** button to delete existing tag
- [x] `onMouseUp` outside of existing tag while editing to submit *without revision*

### Layout
```html
<tag-container>
	<tag>    // submitted
		<span class='text visible'>{tag text}</span>
		<input class='input hidden' />
		<datalist>...</datalist>
		<button class='delete'>X</button>
	</tag>
	<tag>     // active editing
		<span class='text hidden'>{tag text}</span>
		<input class='input visible' />
		<datalist>...</datalist>
		<button class='delete'>X</button>
	</tag>
	<input class='new-tag-input' />
	<datalist>...</datalist>
</tag-container>
```
Does there need to be 'undo' and 'submit' buttons when editing tags?  I don't commonly see it, and I think the keyboard interaction is intuitive enough.  moving your hand to the mouse to click 'submit' after each tag seems really cumbersome.

Regarding datalist:  [mantine.dev](https://mantine.dev/core/tags-input/#group-options) component library has a nice looking tag input, which features 'suggested tags' that can also be grouped.  We can possibly incorporate our existing combobox component to do this.

### Validation

- [x] allow `&`
- [x] `.trim()` input
- [x] case-insensitive entry, but submit as lowercase   _(thoughts?)_

### Props

- [x] label
- [x] values
- [x] valuePatterns
- [x] validators
- [x] placeholder
- [x] notes
- [x] unique
- [x] cannotEdit
- [x] datalist options
- [x] onChange

### Preview
Here is what I have so far.  Not much styling yet.
<img width="409" alt="image" src="https://github.com/naturalcrit/homebrewery/assets/58999374/07e93303-c4b0-4d21-9ad7-661a0024166f"><br />

<img width="409" alt="image" src="https://github.com/naturalcrit/homebrewery/assets/58999374/6cbd992d-74fe-4f99-87bf-9aa409026cf1"><br />

<img width="409" alt="image" src="https://github.com/naturalcrit/homebrewery/assets/58999374/5fab71f2-e4df-40f5-8b0c-dd5312563035"><br />

<img width="409" alt="image" src="https://github.com/naturalcrit/homebrewery/assets/58999374/5bb04caa-ce3c-46e2-b720-f98121a14bc8"><br />

**Comment by ericscheid** [comment]
> case-insensitive entry, but submit as lowercase (thoughts?)

For tag-schemes only. 

We already force-case tag-schemes elswhere (userpage, for classname reasons).

There is value in tags themselves retaining cases when displayed (though should be treated case-insensitively when searching etc)

**Reply by Gazook89** [reply]
so just to be clear, `tag-schemes` = the part before the colon?

**Reply by dbolack-ab** [reply]
The question of Undo and submit falls into usage philosophy for me. If every tag submission is a doc update, then no undo or submit. If the document tags are not updated until a "mass" save, I would consider an undo option

**Comment by 5e-Cleric** [comment]
Could we, maybe possibly consider the dumb idea of.. idk, changing the name of the file to tagInput.jsx? way easier to understand.

**Reply by ericscheid** [reply]
This component is also used for entering invited authors.

**Reply by Gazook89** [reply]
I support tagInput.jsx as well. From my searching on the web, it seems to be a common name for this type of input component. StringArrayEditor is a little unwieldy

**Comment by 5e-Cleric** [comment]
also remember #3284
